src/Singer/Stream.hs

```diff
@@ -49,35 +51,73 @@ toTweet = S.mapMaybe fromTweetSrc
 baseFilter :: (Monad m) => Text -> S.Stream (S.Of TweetSrc) m r -> S.Stream (S.Of TweetSrc) m r
 baseFilter screenName = S.filter (validTweet screenName)
 
-foundParent :: Maybe Tweet -> Tweet -> Maybe Tweet
-foundParent child' parent = pure $ parent { child = child' }
-
-didNotFindParent :: Maybe Tweet -> Tweet -> Maybe Tweet
-didNotFindParent child' parent = let
-  ongoingThread = void . replyTo
-  verifiedChild = do
-    t <- child'
-    ongoingThread t
-    child'
-  in verifiedChild <|> pure parent
-
-extractThread :: Maybe Tweet -> Maybe Thread
-extractThread tweet = do
-  t <- tweet
-  completedThread t
-  hasMultipleTweets t
-  mkThread t
-    where
-      completedThread   = guard . isNothing . replyTo
-      hasMultipleTweets = void . child
-
-
-handleTweet :: Maybe Tweet -> Tweet -> Maybe Tweet
-handleTweet child' parent = let
-  childOfParent = pure (id_ parent) == (child' >>= replyTo)
-  in if childOfParent
-        then foundParent child' parent
-        else didNotFindParent child' parent
+
+data ThreadState = ThreadState
+     { nextTweet     :: Tweet
+     , ongoingThread :: Map TweetId [Tweet]
+     } deriving (Show)
+
+
+handleTweet :: Maybe ThreadState -> Tweet -> Maybe ThreadState
+handleTweet st tw = let
+
+  replyId :: Maybe TweetId
+  replyId = replyTo tw
+
+  getAndDelete :: Ord k => k -> Map k v -> (Maybe v, Map k v)
+  getAndDelete = M.updateLookupWithKey (const $ const Nothing)
```

Comment:
Well since you abstracted `Nothing` to `mempty` in the other module, we should probably do it here too.

Reply:
Indeed, I miss this function in `base`, I wrote it several times, unfortunately.